apis/v1alpha1/opampbridge_webhook_test.go

```diff
@@ -0,0 +1,212 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestOpAMPBridgeDefaultingWebhook(t *testing.T) {
+	one := int32(1)
+	five := int32(5)
+
+	tests := []struct {
+		name        string
+		opampBridge OpAMPBridge
+		expected    OpAMPBridge
+	}{
+		{
+			name:        "provide only required values in spec",
+			opampBridge: OpAMPBridge{},
+			expected: OpAMPBridge{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app.kubernetes.io/managed-by": "opentelemetry-operator",
+					},
+				},
+				Spec: OpAMPBridgeSpec{
+					Replicas:        &one,
+					UpgradeStrategy: UpgradeStrategyAutomatic,
+				},
+			},
+		},
+		{
+			name: "provided optional values in spec",
+			opampBridge: OpAMPBridge{
+				Spec: OpAMPBridgeSpec{
+					Replicas:        &five,
+					UpgradeStrategy: "adhoc",
+				},
+			},
+			expected: OpAMPBridge{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app.kubernetes.io/managed-by": "opentelemetry-operator",
+					},
+				},
+				Spec: OpAMPBridgeSpec{
+					Replicas:        &five,
+					UpgradeStrategy: "adhoc",
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			test.opampBridge.Default()
+			assert.Equal(t, test.expected, test.opampBridge)
+		})
+	}
+}
+
+func TestOpAMPBridgeValidatingWebhook(t *testing.T) {
+
+	two := int32(2)
+
+	tests := []struct { //nolint:govet
+		name        string
+		opampBridge OpAMPBridge
+		expectedErr string
+	}{
+		{
+			name: "specify all required fields, should not return error",
+			opampBridge: OpAMPBridge{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test",
+					Namespace: "default",
+				},
+				Spec: OpAMPBridgeSpec{
+					Endpoint:     "ws://opamp-server:4320/v1/opamp",
+					Protocol:     "wss",
```

Comment:
What does this mean? Can one have `ws://opamp-server:4320/v1/opamp` and protocol as `http`? Another good reason not to have `Protocol` and just rely on the endpoint scheme like the rest of the opamp-go, extension does.